### ISSUE TITLE
Study Graph model API of existing libraries

### ISSUE THREAD

**Issue by DavidBruant** [medialab/graphlib#5]

**Comment by Yomguithereal** [comment]
Adding some references:

* [Cytoscape.js](http://js.cytoscape.org/)
* [Graph](https://www.npmjs.com/package/Graph)
* [graph](https://www.npmjs.com/package/graph)

**Comment by jacomyal** [comment]
From [https://github.com/anvaka/graph-drawing-libraries](https://github.com/anvaka/graph-drawing-libraries):

> * [almende/vis](https://github.com/almende/vis) - Dynamic, browser-based visualization library http://visjs.org
> * [anvaka/ngraph](https://github.com/anvaka/ngraph) - Beautiful Graphs
> * [anvaka/VivaGraphJS](https://github.com/anvaka/VivaGraphJS) - Graph drawing library for JavaScript
> * [cpettitt/dagre](https://github.com/cpettitt/dagre) - Directed graph renderer for javascript
> * [cytoscape/cytoscape.js](https://github.com/cytoscape/cytoscape.js) - An open-source JavaScript graph theory library for analysis and visualisation (compatible with CommonJS/Node.js, AMD/Require.js, jQuery 1.4+, Bower, npm, and plain JavaScript) http://cytoscape.github.com/cytoscape.js
> * [dhotson/springy](https://github.com/dhotson/springy) - A force directed graph layout algorithm in JavaScript http://getspringy.com
> * [fkling/JSNetworkX](https://github.com/fkling/JSNetworkX) - A port of the graph library NetworkX to JavaScript http://jsnetworkx.org
> * [GraphAlchemist/Alchemy](https://github.com/GraphAlchemist/Alchemy) - a graph drawing application built in d3.
> * [jacomyal/sigma.js](https://github.com/jacomyal/sigma.js) - A JavaScript library dedicated to graph drawing http://sigmajs.org
> * [Linkurious/linkurious.js](https://github.com/Linkurious/linkurious.js) - A fork of `jacomyal/sigma.js` bundled with 30+ plugins.
> * [mbostock/d3](https://github.com/mbostock/d3) - A JavaScript visualization library for HTML and SVG. http://d3js.org
> * [samizdatco/arbor](https://github.com/samizdatco/arbor) - a graph visualization library using web workers and jQuery http://arborjs.org
> * [strathausen/dracula](https://github.com/strathausen/dracula) - JavaScript browser based layout and representation of connected graphs. http://www.graphdracula.net

**Comment by Yomguithereal** [comment]
[graphlib](https://github.com/cpettitt/graphlib/wiki)

**Comment by Yomguithereal** [comment]
@jacomyma do you have a link to the Java files defining the Graph API in Gephi?

**Comment by jacomyma** [comment]
I hereby summon master @mbastian (Gephi lead dev). I think that different layers of Gephi are relevant to us right now. A starting point is https://github.com/gephi/gephi/tree/master/modules/GraphAPI but it's not very readable.

**Comment by Yomguithereal** [comment]
Welcome @mbastian!

**Comment by jacomyma** [comment]
...ah but everything's there of course: https://github.com/gephi/graphstore

**Comment by mbastian** [comment]
Here's the Javadoc for GraphStore, Gephi's Graph API: https://gephi.org/graphstore/apidocs/index.html

The two entry points are [GraphModel](https://gephi.org/graphstore/apidocs/index.html?org/gephi/graph/api/GraphModel.html) and [Graph](https://gephi.org/graphstore/apidocs/index.html?org/gephi/graph/api/Graph.html)

Sources are here: https://github.com/gephi/graphstore

You have to make a certain number of choices early on, notably I would say:

1. Decide which type of graphs you want to support is probably the most important. Adding new types afterwards is always very hard.
  - Direction: Directed, undirected, mixed (supports edges of both types in the same graph)
  - Properties: Associate attributes to elements. Decide whether the types are enforced for a given key.
  - Parallel edges: Allow parallel edges between nodes
  - Types edges: Set a specific "type" or "label" to each edge. Encode a specific relationship to an edge.
  - Self-loops: They are annoying as it modifies your degree function
2. Decide how you represent your adjacency list. You have multiple options. A list is poor because removing an edge has O(e) complexity. A map for every node is expensive but easy to implement. The solution Gephi and Neo4j uses is the double-linked-list pointers where an underlying array stores all edges and each edge is linked to others in/out from the same node. Some details [here](https://gephi.wordpress.com/2013/03/05/rebuilding-gephis-core-for-the-0-9-version/). It's the most efficient way that has the best memory/cpu tradeoff. I can elaborate if you're interested.
3. How do you enforce consistency of your graph. That's very important when elements are deleted. Do you take care of deleting all edges attached to a node when deleted for instance? Obviously I would encourage to always go through an API and put private everything so consistency is guaranteed by the API.
4. Can you support edge lookup based on source/target? Some algorithms always need that. The time complexity of doing such a lookup can matter. In Gephi for instance we create a unique long number for each edge so that lookup is O(1). If you support parallel edges that gets immediately more complicated but it can't be missed.

**Comment by Yomguithereal** [comment]
Very important points indeed.

> Direction: Directed, undirected, mixed (supports edges of both types in the same graph)

I guess we should go with mixed as default and indicate directedness though options and/or subclasses if needed later.

> Properties: Associate attributes to elements. Decide whether the types are enforced for a given key.

Apart from the `id` being either a string or a number, I guess we don't have anything else to enforce.

> Parallel edges: Allow parallel edges between nodes

I guess we should support those. This means we should be careful to have getters on edges return arrays everytime not to fall in cases where you have to understand whether the API returned a single edge or multiple ones in a list.

> Self-loops: They are annoying as it modifies your degree function.

Yup. Most library I saw handle this with a boolean flag in the degree function. How do you handle the problem with Gephi?

**Comment by Yomguithereal** [comment]
> How do you enforce consistency of your graph. That's very important when elements are deleted. Do you take care of deleting all edges attached to a node when deleted for instance? Obviously I would encourage to always go through an API and put private everything so consistency is guaranteed by the API.

Yes, the user should use the API to edit the graph structure. If he fails to do so, we shouldn't be accountable for their misfortune :).

Deleting linked edges when deleting a node seems to be the way to go, no? Are there reasons why you would not do so?

**Comment by Yomguithereal** [comment]
> Can you support edge lookup based on source/target? Some algorithms always need that. The time complexity of doing such a lookup can matter. In Gephi for instance we create a unique long number for each edge so that lookup is O(1). If you support parallel edges that gets immediately more complicated but it can't be missed.

It would be nice indeed. We should probably discuss, however, which indexes are opt-out and which are opt-in.

**Comment by jacomyma** [comment]
> Deleting linked edges when deleting a node seems to be the way to go, no? Are there reasons why you would not do so?
For a beginner user, someone using Gephi, GraphCommons or other network softwares, deleting a node implies deleting its edges, because that's what happens visually. I think it is the main point. For a computer scientist the other option (keeping edges) makes sense though, has some algo applications, and I met people finding it better, but they are a very small minority. I don't think they are our target.

**Comment by Yomguithereal** [comment]
There is also a third way (but I just state it here for reference because we shouldn't go this way) which is the Neo4j one: they will yell at you if you delete a node that still has links in the graph. This means that you have to drop all relevant edges explicitly before attempting to remove a node.

**Comment by mbastian** [comment]
> Deleting linked edges when deleting a node seems to be the way to go, no? Are there reasons why you would not do so?

From my experience this is the preferred way, I agree with Mathieu.

> Yup. Most libraries I saw handle this with a boolean flag in the degree function. How do you handle the problem with Gephi?

Sorry I mispoke the degree function can remain `indegree+outdegree` for self-loops - this is what the theory says. So a node with self-loop has a degree of 2 for both the directed and undirected case in Gephi. The special case is with edge iteration where you expect the edge to be returned only once when you ask the edges for a given node.

> I guess we should go with mixed as default and indicate directedness though options and/or subclasses if needed later.

The mixed case makes things slightly complicated for functions such as `getEdge(source, target)` as if an undirected edge exists between these two nodes `getEdge(source, target)` should be equal to `getEdge(target, source)`. If you enforce directed or undirected for the entire graph this lookup is easier.
